pkg/tagging/base.go

```diff
@@ -0,0 +1,127 @@
+}
+
+// NewCmdUpdateTag creates a command object for the command
+func NewCmdUpdateTag(tagVerb, tagType string) (*cobra.Command, *Options) {
+	o := &Options{}
+	caser := cases.Title(language.English)
+	cmd := &cobra.Command{
+		Use:     tagVerb,
+		Short:   fmt.Sprintf("%ss all kubernetes resources in the given directory tree", caser.String(tagVerb)),
+		Long:    fmt.Sprintf(tagLong, caser.String(tagVerb)),
+		Example: fmt.Sprintf(tagExample, tagVerb, rootcmd.BinaryName, tagType),
+		Run: func(cmd *cobra.Command, args []string) {
+			err := UpdateTagInYamlFiles(o.Dir, tagType+"s", args, o.Filter, o.PodSpec, o.Overwrite)
+			helper.CheckErr(err)
+		},
+	}
+	cmd.Flags().StringVarP(&o.Dir, "dir", "", ".", "the directory to recursively look for the *.yaml or *.yml files")
+	cmd.Flags().BoolVarP(&o.PodSpec, "pod-spec", "p", false,
+		fmt.Sprintf("%s the PodSpec in spec.template.metadata.%ss (or spec.jobTemplate.spec.template.metadata.%[2]ss for CronJobs) rather than the top level %[2]ss", tagVerb, tagType))
+	cmd.Flags().BoolVar(&o.Overwrite, "overwrite", true, "Set to false to not overwrite any existing value")
+	o.Filter.AddFlags(cmd)
+	return cmd, o
+}
+
+// UpdateTagInYamlFiles updates the annotations in yaml files
+func UpdateTagInYamlFiles(dir string, tagType string, tags []string, filter kyamls.Filter, podSpec bool, override bool) error { //nolint:gocritic
+	modifyFn := func(node *yaml.RNode, path string) (bool, error) {
+		sort.Strings(tags)
+		pathArray := []string{"metadata", tagType}
+		if podSpec {
+			pathArray = append([]string{"spec", "template"}, pathArray...)
+			if kyamls.GetKind(node, path) == "CronJob" {
+				pathArray = append([]string{"spec", "jobTemplate"}, pathArray...)
+			}
+		}
+		pathNode, err := node.Pipe(yaml.PathGetter{Path: pathArray, Create: yaml.MappingNode})
+		if err != nil {
+			return false, errors.Wrapf(err, "failed to get %v", pathArray)
+		}
+
+		var modified bool
+		for _, a := range tags {
+			paths := strings.SplitN(a, "=", 2)
+			k := paths[0]
+			v := ""
+			if len(paths) > 1 {
+				v = paths[1]
+			} else if strings.HasSuffix(k, "-") {
+				rNode, err := pathNode.Pipe(yaml.Clear(strings.TrimSuffix(k, "-")))
+				if err != nil {
+					return modified, err
+				}
+				modified = rNode != nil
+				continue
+			}
+			vn := yaml.NewScalarRNode(v)
+			vn.YNode().Tag = yaml.NodeTagString
+			vn.YNode().Style = yaml.SingleQuotedStyle
+
+			field, err := pathNode.Pipe(yaml.FieldMatcher{Name: k})
+			if err != nil {
+				return modified, errors.Wrapf(err, "failed to match %s", k)
+			}
+			if field != nil {
+				if !override {
+					continue
+				}
+				// need to def ref the Node since field is ephemeral
+				field.SetYNode(vn.YNode())
+				modified = true
+			} else {
+				// create the field
+				pathNode.YNode().Content = append(
+					pathNode.Content(),
+					&yaml.Node{
+						Kind:  yaml.ScalarNode,
+						Value: k,
+					},
+					vn.YNode())
+				modified = true
+			}
+		}
+		return modified, nil
+	}
```

Comment:
this is suuuper high cognitive complexity, we should tyr to split it up a little I think so it's more understandable

Reply:
Well, that's a matter of opinion... But I'll split it up a bit